### ISSUE TITLE
C functions that can trigger GC need special handling

### ISSUE THREAD

**Issue by bjourne** [slavapestov/factor#1468]
Say you have a function:
```factor
FUNCTION: int do_stuff ( int x, Rect* r )
```
Which you call using `123 0 0 50 50 Rect <struct-boa> do_stuff`. It works in 99% of the cases. But if `do_stuff` somehow activates a callback and runs Factor code, then it can trigger a gc and the pointer to the Rect struct will in the C function be invalid. The problem is described in the manual: http://docs.factorcode.org/content/article-byte-arrays-gc.html

I committed a fix for a function call that triggered the above annoyance (5684fab4841eae5553626764354111fdb2383bd3). But there are many more functions that runs callbacks which might have the same problem. So instead of fixing them one by one with special code it would be nice to have some syntax, like:
```factor
SAFE-FUNCTION: GpStatus GdipBitmapLockBits ( GpBitmap* x, GpRect* x, UINT x,
             PixelFormat x, BitmapData* x )
```
`SAFE-FUNCTION:` would mean it copies all pointers to the callstack before calling the function. Something like that.

**Comment by AlexIljin** [comment]
What would copying all pointers to the callstack do to mitigate the issue? Are pointers on the callstack somehow exempt from relocation by the GC?

**Comment by bjourne** [comment]
Exactly. Objects allocated in the stack are never moved. But I don't know if that is the best solution. Other runtimes solve the same problem by using [pinned pointers](https://msdn.microsoft.com/en-us/library/aa712862(v=vs.71).aspx) or by copying objects to a safe memory area like JNI does.

**Comment by AlexIljin** [comment]
What if the external function returns immediately, but performs some asynchronous operation on the data pointed to by the passed pointer? If I understand the situation correctly, we must use special memory allocation for that. Is there a better solution possible? Or am I thinking of a completely different use case now?

**Comment by bjourne** [comment]
Fortunately, that can't happen as Factor doesn't (yet) use real/hardware threads. :P You are correct that that situation would be incredibly tricky for Factor to handle because from its perspective, the pointer would be lost. It can't know that C code running asynchronously still references it. 

One can kind of solve the problem by copying data from the heap into an immovable space. But that adds significant overhead to each ffi call which is not fun. Perhaps there is no good solution and you can't get both safety and speed.
